### ISSUE TITLE
Can't upgrade from 0.3.2 to 0.4.1 or later due to "unbound type variable: T"

### ISSUE THREAD

**Issue by samiojanpaa** [EvidentSolutions/apina#31]
`FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':citizen-web-api:apina'.
> unbound type variable: T

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':citizen-web-api:apina'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:69)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.PostExecutionAnalysisTaskExecuter.execute(PostExecutionAnalysisTaskExecuter.java:35)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:64)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:53)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:233)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:215)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.processTask(AbstractTaskPlanExecutor.java:74)
        at org.gradle.execution.taskgraph.AbstractTaskPlanExecutor$TaskExecutorWorker.run(AbstractTaskPlanExecutor.java:55)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor.process(DefaultTaskPlanExecutor.java:32)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter.execute(DefaultTaskGraphExecuter.java:113)
        at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.access$000(DefaultBuildExecuter.java:23)
        at org.gradle.execution.DefaultBuildExecuter$1.proceed(DefaultBuildExecuter.java:43)
        at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:32)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:37)
        at org.gradle.execution.DefaultBuildExecuter.execute(DefaultBuildExecuter.java:30)
        at org.gradle.initialization.DefaultGradleLauncher$4.run(DefaultGradleLauncher.java:186)
        at org.gradle.internal.Factories$1.create(Factories.java:22)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:53)
        at org.gradle.initialization.DefaultGradleLauncher.doBuildStages(DefaultGradleLauncher.java:183)
        at org.gradle.initialization.DefaultGradleLauncher.access$200(DefaultGradleLauncher.java:33)
        at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:112)
        at org.gradle.initialization.DefaultGradleLauncher$1.create(DefaultGradleLauncher.java:106)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:91)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:63)
        at org.gradle.initialization.DefaultGradleLauncher.doBuild(DefaultGradleLauncher.java:106)
        at org.gradle.initialization.DefaultGradleLauncher.run(DefaultGradleLauncher.java:92)
        at org.gradle.launcher.exec.GradleBuildController.run(GradleBuildController.java:66)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:28)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:41)
        at org.gradle.launcher.exec.InProcessBuildActionExecuter.execute(InProcessBuildActionExecuter.java:26)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:79)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecuter.execute(ContinuousBuildActionExecuter.java:51)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:59)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:47)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:26)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:34)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:74)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.call(ForwardClientInput.java:72)
        at org.gradle.util.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:72)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:60)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:72)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:36)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.HintGCAfterBuild.execute(HintGCAfterBuild.java:44)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:120)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:50)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:293)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
        at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
Caused by: java.lang.RuntimeException: unbound type variable: T
        at fi.evident.apina.java.model.type.JavaTypeVariable.lambda$resolve$9(JavaTypeVariable.java:29)
        at fi.evident.apina.java.model.type.JavaTypeVariable.resolve(JavaTypeVariable.java:29)
        at fi.evident.apina.java.model.type.TypeEnvironment.lambda$resolve$8(TypeEnvironment.java:37)
        at fi.evident.apina.utils.CollectionUtils.map(CollectionUtils.java:15)
        at fi.evident.apina.java.model.type.TypeEnvironment.resolve(TypeEnvironment.java:37)
        at fi.evident.apina.spring.JacksonTypeTranslator.lambda$boundClassFor$48(JacksonTypeTranslator.java:264)
        at fi.evident.apina.spring.JacksonTypeTranslator.boundClassFor(JacksonTypeTranslator.java:261)
        at fi.evident.apina.spring.JacksonTypeTranslator.classesUpwardsFrom(JacksonTypeTranslator.java:241)
        at fi.evident.apina.spring.JacksonTypeTranslator.ignoredProperties(JacksonTypeTranslator.java:206)
        at fi.evident.apina.spring.JacksonTypeTranslator.initClassDefinition(JacksonTypeTranslator.java:173)
        at fi.evident.apina.spring.JacksonTypeTranslator.translateClassType(JacksonTypeTranslator.java:150)
        at fi.evident.apina.spring.JacksonTypeTranslator.access$200(JacksonTypeTranslator.java:25)
        at fi.evident.apina.spring.JacksonTypeTranslator$1.visit(JacksonTypeTranslator.java:84)
        at fi.evident.apina.spring.JacksonTypeTranslator$1.visit(JacksonTypeTranslator.java:51)
        at fi.evident.apina.java.model.type.JavaBasicType.accept(JavaBasicType.java:39)
        at fi.evident.apina.spring.JacksonTypeTranslator.translateType(JacksonTypeTranslator.java:51)
        at fi.evident.apina.spring.JacksonTypeTranslator$1.visit(JacksonTypeTranslator.java:100)
        at fi.evident.apina.spring.JacksonTypeTranslator$1.visit(JacksonTypeTranslator.java:51)
        at fi.evident.apina.java.model.type.JavaParameterizedType.accept(JavaParameterizedType.java:43)
        at fi.evident.apina.spring.JacksonTypeTranslator.translateType(JacksonTypeTranslator.java:51)
        at fi.evident.apina.spring.SpringModelReader.resolveResponseBody(SpringModelReader.java:108)
        at fi.evident.apina.spring.SpringModelReader.createEndpointForMethod(SpringModelReader.java:70)
        at fi.evident.apina.spring.SpringModelReader.createEndpointGroupForController(SpringModelReader.java:64)
        at fi.evident.apina.spring.SpringModelReader.createEndpointsForControllers(SpringModelReader.java:55)
        at fi.evident.apina.spring.SpringModelReader.readApiDefinition(SpringModelReader.java:48)
        at fi.evident.apina.ApinaProcessor.process(ApinaProcessor.java:26)
        at fi.evident.apina.gradle.tasks.ApinaTask.generateTypeScript(ApinaTask.java:54)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:75)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.doExecute(DefaultTaskClassInfoStore.java:136)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:129)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:118)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:623)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:606)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:80)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:61)
        ... 68 more
`

**Comment by komu** [comment]
I have some code that will probably fix this already sitting in a branch, but I'd like to be sure that I understand your specific problem fully. So could you provide a minimal test-case that reproduces the problem? Delete controllers, methods and Jackson classes until only a couple of classes remain. I'm pretty sure that the problem can be reproduced in 10 lines of code.

**Comment by samiojanpaa** [comment]
Controller that causes the problem:

```
package fi.my.controller;

import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class MyController {

    @RequestMapping("/api/my_get")
    @ResponseBody
    public ResponseEntity<byte[]> getMyResponse() {
        return new ResponseEntity<>(new byte[1], new HttpHeaders(), HttpStatus.OK);
    }

}

```

Should apina be able to handle that?

**Comment by komu** [comment]
Yeah, seems legit. I'll look into that. Thanks!

**Comment by komu** [comment]
Actually I can't reproduce your exception using Apina 0.7.0.

What version are you running? You mention "0.4.1 or later" and based on stacktrace it's quite clear that at least it's not 0.7.0. So is it 0.4.1? Please test against 0.7.0, you'll need to specify `framework = 'angular1'`, but otherwise things should be compatible.

There's a problem though. Apina does not have special handling for `ResponseEntity` or `HttpEntity` and does not realize that the effective return type is actually `byte[]`. However, this causes no exception, but an incorrect signature for the endpoint. I'll look into this.

**Comment by komu** [comment]
I just released [0.7.1](https://github.com/EvidentSolutions/apina/releases/tag/v0.7.1) that works with your test case and produces correct code for it.

I'll close this now, if you still have problems on 0.7.1, reopen or file a new issue.
